### ISSUE TITLE
New Lint: Use rest pattern instead of ignoring values when destructuring structs and enums

### ISSUE THREAD

**Issue by wainwrightmark** [rust-lang/rust-clippy#10666]
### What it does

This lint will use a rest pattern (`..`), instead of ignoring values with `_` when destructuring structs and enums. 

I'm suggesting to only do this for `_`, and leave as they are the cases where the values are named and prefixed with an underscore.

### Lint Name

ignored_field_when_destructuring

### Category

style

### Advantage

- This reduces noise and makes it more obvious which fields are relevant, especially when there are many ignored values.

### Drawbacks

- This will prevent a compilation error if additional fields are added to the struct. Some programmers may want that error to remind them to do something with the newly added fields. 

### Example

```rust
struct Point {
    x: i32,
    y: i32,
}

let origin = Point { x: 0, y: 0 };
let Point { x, y: _ } = origin;

println!("x is {}", x);
```

Could be written as:

```rust
struct Point {
    x: i32,
    y: i32,
}

let origin = Point { x: 0, y: 0 };
let Point { x, .. } = origin;

println!("x is {}", x);
```

**Comment by Alexendoo** [comment]
That drawback is a pretty big one, I would say this would have to be `restriction` if it were added

**Comment by wainwrightmark** [comment]
> That drawback is a pretty big one, I would say this would have to be `restriction` if it were added

Yes, I think I undervalued this originally. I'm sure there's at least a few people who have written code like this with the deliberate intent of getting compiler errors if a field is added and we definitely shouldn't break that.

On the other hand I've realised why my code is full of ugly expressions that I'd like to fix with this lint. Rust Analyzer's `fill match arms` generates code like `Foo::Bar{a,b,c} => todo!()` and then if you don't use `b` or `c`, clippy turns it into something like `Foo::Bar{a,b: _,c: _} => do_something(a)`
So I imagine I'm not the only one who would appreciate being able to opt into a lint which tidies this up.

**Comment by xxchan** [comment]
Actually I want the reverse of this lint: ignore values instead of rest pattern. Because I found in our codebase, the struct fields are omitted by mistake, so I want to make it explicit what fields are ignored.
